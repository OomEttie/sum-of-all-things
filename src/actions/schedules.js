const uuidV4 = require('uuid/v4');
import database, { schedulesDatabaseRef } from '../utils/firebase';

import {
  ACTION_SCHEDULE_ADD,
  ACTION_SCHEDULE_EDIT,
  ACTION_SCHEDULE_LIST,
  ACTION_SCHEDULE_REMOVE
} from '../actions/actions-list';

//
// action schedule add
//
export const addSchedule = schedule => ({
  type: ACTION_SCHEDULE_ADD,
  schedule
});

export const startAddSchedule = (scheduleData = {}) => {
  return (dispatch, getState) => {
    const {
      description = '',
      start = 0,
      end = 0,
      client = '',
      note = ''
    } = scheduleData;

    const schedule = { description, start, end, client, note };

    const dbPath = schedulesDatabaseRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );

    return database
      .ref(dbPath)
      .push(schedule)
      .then(ref => {
        dispatch(
          addSchedule({
            id: ref.key,
            ...schedule
          })
        );
      });
  };
};

//
// remove schedule
//
export const removeSchedule = ({ id } = {}) => ({
  type: ACTION_SCHEDULE_REMOVE,
  id: id
});

export const startRemoveSchedule = ({ id }) => {
  return (dispatch, getState) => {
    const dbPath = schedulesDatabaseRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );

    return database
      .ref(`${dbPath}/${id}`)
      .remove()
      .then(data => {
        dispatch(removeSchedule({ id }));
      });
  };
};

//
// EDIT_CLIENT
//
export const editSchedule = (id, updates) => ({
  type: ACTION_SCHEDULE_EDIT,
  id,
  updates
});

export const startEditSchedule = (id, updates) => {
  return (dispatch, getState) => {
    let dbPath = schedulesDatabaseRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );

    return database
      .ref(`${dbPath}/${id}`)
      .update({
        ...updates
      })
      .then(() => {
        dispatch(editSchedule(id, updates));
      });
  };
};

//
// list schedules
//
export const listSchedules = schedules => ({
  type: ACTION_SCHEDULE_LIST,
  schedules
});

export const startListSchedules = () => {
  return (dispatch, getState) => {
    const dbPath = schedulesDatabaseRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    return database
      .ref(dbPath)
      .once('value')
      .then(snapshot => {
        const schedulesData = [];
        snapshot.forEach(snapshotChild => {
          schedulesData.push({
            id: snapshotChild.key,
            ...snapshotChild.val()
          });
        });
        dispatch(listSchedules(schedulesData));
      })
      .catch(e => {});
  };
};
