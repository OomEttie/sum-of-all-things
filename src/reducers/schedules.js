import {
  ACTION_SCHEDULE_ADD,
  ACTION_SCHEDULE_REMOVE,
  ACTION_SCHEDULE_EDIT,
  ACTION_SCHEDULE_LIST
} from '../actions/actions-list';

const schedulesReducerDefaultState = [];

export const schedulesReducer = (
  state = schedulesReducerDefaultState,
  action
) => {
  switch (action.type) {
    case ACTION_SCHEDULE_ADD:
      return [...state, action.schedule];
    case ACTION_SCHEDULE_REMOVE:
      return state.filter(data => data.id != action.id);
    case ACTION_SCHEDULE_EDIT: {
      return state.map(schedule => {
        if (schedule.id == action.id) {
          return { ...schedule, ...action.updates };
        } else {
          return schedule;
        }
      });
    }
    case ACTION_SCHEDULE_LIST:
      return action.schedules;
    default:
      return state;
  }
};
