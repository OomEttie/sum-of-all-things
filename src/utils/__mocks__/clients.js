export const clientObject = (id, name, surname, note) => {
  return {
    id: id,
    name: name,
    surname: surname,
    note: note  
  };
}

const client1 = clientObject('1', '1 name', '1 surname', '1 note');
const client2 = clientObject('2', '2 name', '2 surname', '2 note');
const client3 = clientObject('3', '3 name', '3 surname', '3 note');

export const clients = [
  client1,
  client2,
  client3
];

