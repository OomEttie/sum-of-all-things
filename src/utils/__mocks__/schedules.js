export const scheduleObject = (id, description, start, end, client, note) => {
  return {
    id: id,
    description: description,
    start: start,
    end: end,
    client: client,
    note: note
  };
};

const schedule1 = scheduleObject('1', 'schedule 1', 1533117600, 1533117600, 'client 1', 'note 1');
const schedule2 = scheduleObject('2', 'schedule 2', 1533117600, 1533117600, 'client 2', 'note 2');
const schedule3 = scheduleObject('3', 'schedule 3', 1533117600, 1533117600, 'client 3', 'note 3');

export const schedules = [schedule1, schedule2, schedule3];
