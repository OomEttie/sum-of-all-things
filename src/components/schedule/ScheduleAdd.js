import React from 'react';
import { connect } from 'react-redux';
import { ScheduleForm } from './ScheduleForm';
import { startAddSchedule } from '../../actions/schedules';
import { schedules } from '../../utils/__mocks__/schedules';

export class ScheduleAdd extends React.Component {
  onSubmit = schedule => {
    this.props.addSchedule(schedule);
    this.props.history.push('/schedule');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Schedule</h1>
          </div>
        </div>
        <div className="content-container">
          <ScheduleForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    addSchedule: schedule => {
      dispatch(startAddSchedule(schedule));
    }
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(ScheduleAdd);
