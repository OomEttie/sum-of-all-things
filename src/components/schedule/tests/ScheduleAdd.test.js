import React from 'react';
import { shallow } from 'enzyme';
import { ScheduleAdd } from '../ScheduleAdd';
import { schedules } from '../../../utils/__mocks__/schedules';

let addSchedule, history, wrapper;

beforeEach(() => {
  addSchedule = jest.fn();
  history = { push: jest.fn() };

  wrapper = shallow(<ScheduleAdd addSchedule={addSchedule} history={history} />);
});

test('render add schedule page correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('handle add schedule onSubmit event', () => {
  wrapper.find('ScheduleForm').prop('onSubmit')(schedules[1]);

  expect(history.push).toHaveBeenLastCalledWith('/schedule');
  expect(addSchedule).toHaveBeenLastCalledWith(schedules[1]);
});
