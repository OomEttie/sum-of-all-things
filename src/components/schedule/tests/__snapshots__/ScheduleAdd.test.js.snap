// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render add schedule page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Schedule
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ScheduleForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
